File: plugins/skill-intake/scripts/optimize_layout.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

function parseEdges(src) {
  const edges = [];
  const re = /^\s*([A-Za-z0-9_]+)(?:\[[^\]]*\])?\s*-->\s*(?:\|[^|]*\|\s*)?([A-Za-z0-9_]+)/gm;
  let m;
  while ((m = re.exec(src))) edges.push([m[1], m[2]]);
  return edges;
}

function topoOrder(edges) {
  const nodes = new Set();
  const incoming = new Map();
  const outgoing = new Map();
  for (const [a, b] of edges) {
    nodes.add(a); nodes.add(b);
    incoming.set(b, (incoming.get(b) || 0) + 1);
    if (!outgoing.has(a)) outgoing.set(a, []);
    outgoing.get(a).push(b);
  }
  const order = [];
  const queue = [...nodes].filter((n) => !incoming.get(n));
  const inDeg = new Map([...nodes].map((n) => [n, incoming.get(n) || 0]));
  while (queue.length) {
    const n = queue.shift();
    order.push(n);
    for (const next of outgoing.get(n) || []) {
      inDeg.set(next, inDeg.get(next) - 1);
      if (inDeg.get(next) === 0) queue.push(next);
    }
  }
  return order.length === nodes.size ? order : [...nodes];
}

function optimize(src) {
  const edges = parseEdges(src);
  const order = topoOrder(edges);
  return { order, edge_count: edges.length, suggestion: order.join(' -> ') };
}

function main(argv) {
  const file = argv[2];
  if (!file) { process.stderr.write('usage: optimize_layout.js <diagram.mmd>\n'); return 2; }
  let src;
  try { src = fs.readFileSync(path.resolve(file), 'utf8'); }
  catch (e) { process.stderr.write(`input error: ${e.message}\n`); return 2; }
  const r = optimize(src);
  process.stdout.write(JSON.stringify(r, null, 2) + '\n');
  return 0;
}

module.exports = { optimize, topoOrder };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: plugins/skill-intake/scripts/optimize_layout.py
#!/usr/bin/env python3
"""Suggest a topological order of nodes in a Mermaid flow diagram."""
from __future__ import annotations

import json
import re
import sys
from collections import deque
from pathlib import Path
from typing import Any

_EDGE_RE = re.compile(
    r'^\s*([A-Za-z0-9_]+)(?:\[[^\]]*\])?\s*-->\s*(?:\|[^|]*\|\s*)?([A-Za-z0-9_]+)',
    re.MULTILINE,
)


def parse_edges(src: str) -> list[tuple[str, str]]:
    return [(m.group(1), m.group(2)) for m in _EDGE_RE.finditer(src)]


def topo_order(edges: list[tuple[str, str]]) -> list[str]:
    nodes: set[str] = set()
    incoming: dict[str, int] = {}
    outgoing: dict[str, list[str]] = {}
    for a, b in edges:
        nodes.add(a)
        nodes.add(b)
        incoming[b] = incoming.get(b, 0) + 1
        outgoing.setdefault(a, []).append(b)
    in_deg = {n: incoming.get(n, 0) for n in nodes}
    queue = deque(n for n in nodes if in_deg[n] == 0)
    order: list[str] = []
    while queue:
        n = queue.popleft()
        order.append(n)
        for nxt in outgoing.get(n, []):
            in_deg[nxt] -= 1
            if in_deg[nxt] == 0:
                queue.append(nxt)
    if len(order) != len(nodes):
        return list(nodes)
    return order


def optimize(src: str) -> dict[str, Any]:
    edges = parse_edges(src)
    order = topo_order(edges)
    return {'order': order, 'edge_count': len(edges), 'suggestion': ' -> '.join(order)}


def main(argv: list[str]) -> int:
    if len(argv) < 1:
        sys.stderr.write('usage: optimize_layout.py <diagram.mmd>\n')
        return 2
    try:
        src = Path(argv[0]).resolve().read_text(encoding='utf-8')
    except Exception as e:
        sys.stderr.write(f'input error: {e}\n')
        return 2
    r = optimize(src)
    sys.stdout.write(json.dumps(r, ensure_ascii=False, indent=2) + '\n')
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
